cstar/roms/simulation.py

```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+        - This method overrides runtime settings using values from `self.discretization`,
+          `self.model_grid`, `self.initial_conditions`, and all forcing datasets.
+        - If MARBL configuration files are present in `runtime_code.files`, their paths
+          are also included.
         """
 
-        in_files = []
-        if self.runtime_code is None:
+        if self.runtime_code.working_path is None:
             raise ValueError(
-                "ROMSSimulation.runtime_code not set."
-                + " ROMS requires a runtime options file "
-                + "(typically roms.in)"
+                "Cannot access runtime settings without local "
+                + "`.in` file. Call ROMSSimulation.setup() or "
+                + "ROMSSimulation.runtime_code.get() and try again."
             )
 
-        in_files = [
-            fname.replace(".in_TEMPLATE", ".in")
-            for fname in self.runtime_code.files
-            if (fname.endswith(".in") or fname.endswith(".in_TEMPLATE"))
-        ]
-        if len(in_files) > 1:
-            raise ValueError(
-                "Multiple '.in' files found:"
-                + "\n{in_files}"
-                + "\nROMS runtime file choice ambiguous"
+        simulation_runtime_settings = ROMSRuntimeSettings.from_file(
+            self.runtime_code.working_path / self._in_file
+        )
+
+        # Previous modifications
+        # Time step entry
+        simulation_runtime_settings.time_stepping.dt = self.discretization.time_step
+
+        # ntimesteps entry:
+        simulation_runtime_settings.time_stepping.ntimes = self._n_time_steps
+
+        # Initial conditions entry
+        if self.initial_conditions:
+            simulation_runtime_settings.initial.ininame = (
+                self.initial_conditions.path_for_roms[0]
             )
-        elif len(in_files) == 0:
-            raise ValueError(
-                "No '.in' file found in ROMSSimulation.runtime_code."
-                + "ROMS expects a runtime options file with the '.in'"
-                + "extension, e.g. roms.in"
+
+        # Grid entry
+        if self.model_grid:
+            simulation_runtime_settings.grid = cstar.roms.runtime_settings.Grid(
+                grid=self.model_grid.path_for_roms[0]
             )
         else:
-            if self.runtime_code.working_path is not None:
-                return self.runtime_code.working_path / in_files[0]
-            else:
-                return Path(in_files[0])
+            simulation_runtime_settings.grid = None
+
+        # Forcing
+        simulation_runtime_settings.forcing.filenames = self._forcing_paths
+        # MARBL settings:
+
+        if all(
+            f in self.runtime_code.files
+            for f in [
+                "marbl_in",
+                "marbl_tracer_output_list",
+                "marbl_diagnostic_output_list",
+            ]
+        ):
+            simulation_runtime_settings.marbl_biogeochemistry = (
```

Comment:
this line of code is incomprehensible due to the formatter. this should be broken into individual variable sets isntead of nesting and doing repeated line-continuations.